tasty-query/shared/src/main/scala/tastyquery/Classpaths.scala

```diff
@@ -26,35 +26,74 @@ object Classpaths:
         val lastSep = binaryName.lastIndexOf('.')
         if lastSep == -1 then ("", binaryName)
-        else {
+        else
           import scala.language.unsafeNulls
           val packageName = binaryName.substring(0, lastSep)
           val className = binaryName.substring(lastSep + 1)
           (packageName, className)
+
+      def filterEntry(entry: Classpath.Entry, lookup: Map[String, List[String]]) =
+        val packages = entry.packages.collect {
+          case pkg if lookup.contains(pkg.dotSeparatedName) =>
+            val tastys = pkg.tastys.filter(t => lookup(pkg.dotSeparatedName).contains(t.binaryName))
+            val classes = pkg.classes.filter(c => lookup(pkg.dotSeparatedName).contains(c.binaryName))
+            if tastys.isEmpty && classes.isEmpty then None // perhaps the class came from another entry
+            else Some(PackageData(pkg.dotSeparatedName, classes, tastys))
         }
-      }
+        Classpath.Entry(packages.flatten)
+
       val formatted = binaryNames.map(packageAndClass)
-      val grouped = formatted.groupMap((pkg, _) => pkg)((_, cls) => cls)
-      val filtered = packages.collect {
-        case pkg if grouped.contains(pkg.dotSeparatedName) =>
-          val tastys = pkg.tastys.filter(t => grouped(pkg.dotSeparatedName).contains(t.binaryName))
-          val classes = pkg.classes.filter(c => grouped(pkg.dotSeparatedName).contains(c.binaryName))
-          PackageData(pkg.dotSeparatedName, classes, tastys)
-      }
-      new Classpath(filtered)
+      val lookup = formatted.groupMap((pkg, _) => pkg)((_, cls) => cls)
+      val filtered = entries.map(filterEntry(_, lookup))
+      Classpath(filtered)
     end withFilter
   end Classpath
 
   /** Factory object for [[Classpath]] instances. */
   object Classpath {
 
-    /** Constructs an instance of [[Classpath]] with the given [[PackageData]]s. */
-    def from(packages: IArray[PackageData]): Classpath =
-      new Classpath(packages)
+    /** A [[Classpath.Entry]] encapsulates the package data for a single classpath entry
+      * (i.e. a given directory or jar file).
+      * Can only be created by [[Classpath.from]]. You can lookup all symbols
+      * originating from this entry
+      * with [[Contexts.Context.findSymbolsByClasspathEntry ctx.findSymbolsByClasspathEntry]].
+      * e.g.:
+      * ```scala
+      * val classpath = ClasspathLoaders.read(myLibraryPath :: stdLibPaths)
+      * given Context = Contexts.init(classpath)
+      * val myLibSyms = ctx.findSymbolsByClasspathEntry(classpath.entries.head)
+      * ```
+      */
+    final class Entry private[Classpath] (val packages: IArray[PackageData])
+
+    /** Creates a [[Classpath]] from a sequence of classpath entries. Each entry corresponds to a single directory
+      * or jar file, and represents the various `.class` and `.tasty` files found within it.
+      */
+    def from(entries: IArray[PackageData]*): Classpath =
```

Comment:
```suggestion
    def from(entries: Seq[IArray[PackageData]]): Classpath =
```
?
It seems that all use cases will have a dynamic sequence here, and never a fixed number of elements.

Reply:
fixed